File: presto-main/src/main/java/com/facebook/presto/operator/scalar/JsonPathTokenizer.java
/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.facebook.presto.operator.scalar;

import com.facebook.presto.spi.PrestoException;
import com.google.common.collect.AbstractIterator;

import static com.facebook.presto.spi.StandardErrorCode.INVALID_FUNCTION_ARGUMENT;
import static java.lang.Character.isLetterOrDigit;
import static java.lang.String.format;
import static java.util.Objects.requireNonNull;

public class JsonPathTokenizer
        extends AbstractIterator<String>
{
    private static final char QUOTE = '\"';
    private static final char DOT = '.';
    private static final char OPEN_BRACKET = '[';
    private static final char CLOSE_BRACKET = ']';
    private static final char UNICODE_CARET = '\u2038';

    private final String path;
    private int index;

    public JsonPathTokenizer(String path)
    {
        this.path = requireNonNull(path, "path is null");

        if (path.isEmpty()) {
            throw invalidJsonPath();
        }

        // skip the start token
        match('$');
    }

    @Override
    protected String computeNext()
    {
        if (!hasNextCharacter()) {
            return endOfData();
        }

        if (tryMatch(DOT)) {
            return matchPathSegment();
        }

        if (tryMatch(OPEN_BRACKET)) {
            String token = tryMatch(QUOTE) ? matchQuotedSubscript() : matchUnquotedSubscript();

            match(CLOSE_BRACKET);
            return token;
        }

        throw invalidJsonPath();
    }

    private String matchPathSegment()
    {
        // seek until we see a special character or whitespace
        int start = index;
        while (hasNextCharacter() && isUnquotedPathCharacter(peekCharacter())) {
            nextCharacter();
        }
        int end = index;

        String token = path.substring(start, end);

        // an empty unquoted token is not allowed
        if (token.isEmpty()) {
            throw invalidJsonPath();
        }

        return token;
    }

    private static boolean isUnquotedPathCharacter(char c)
    {
        return c == ':' || isUnquotedSubscriptCharacter(c);
    }

    private String matchUnquotedSubscript()
    {
        // seek until we see a special character or whitespace
        int start = index;
        while (hasNextCharacter() && isUnquotedSubscriptCharacter(peekCharacter())) {
            nextCharacter();
        }
        int end = index;

        String token = path.substring(start, end);

        // an empty unquoted token is not allowed
        if (token.isEmpty()) {
            throw invalidJsonPath();
        }

        return token;
    }

    private static boolean isUnquotedSubscriptCharacter(char c)
    {
        return c == '_' || isLetterOrDigit(c);
    }

    private String matchQuotedSubscript()
    {
        // quote has already been matched

        // seek until we see the close quote
        int start = index;
        while (hasNextCharacter() && peekCharacter() != QUOTE) {
            nextCharacter();
        }
        int end = index;

        String token = path.substring(start, end);

        match(QUOTE);
        return token;
    }

    private boolean hasNextCharacter()
    {
        return index < path.length();
    }

    private void match(char expected)
    {
        if (!tryMatch(expected)) {
            throw invalidJsonPath();
        }
    }

    private boolean tryMatch(char expected)
    {
        if (!hasNextCharacter() || peekCharacter() != expected) {
            return false;
        }
        index++;
        return true;
    }

    private void nextCharacter()
    {
        index++;
    }

    private char peekCharacter()
    {
        return path.charAt(index);
    }

    private PrestoException invalidJsonPath()
    {
        return new PrestoException(INVALID_FUNCTION_ARGUMENT, format("Invalid JSON path: '%s'", path));
    }

    @Override
    public String toString()
    {
        return path.substring(0, index) + UNICODE_CARET + path.substring(index);
    }
}

File: velox/functions/prestosql/json/JsonPathTokenizer.cpp
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "velox/functions/prestosql/json/JsonPathTokenizer.h"

namespace facebook::velox::functions {

namespace {
const char ROOT = '$';
const char DOT = '.';
const char DASH = '-';
const char DOUBLE_QUOTE = '"';
const char SINGLE_QUOTE = '\'';
const char STAR = '*';
const char BACK_SLASH = '\\';
const char UNDER_SCORE = '_';
const char OPEN_BRACKET = '[';
const char CLOSE_BRACKET = ']';
const char AT = '@';
const char OPEN_PARANTHESIS = '(';

bool isUnquotedBracketKeyFormat(char c) {
  return c == UNDER_SCORE || c == STAR || c == DASH || std::isalnum(c);
}

// Whitespaces are not allowed in dot key format. We check for it explicitly
// outside of this function.
bool isDotKeyFormat(char c) {
  return c != OPEN_BRACKET && c != DOT && c != OPEN_PARANTHESIS;
}

} // namespace

bool JsonPathTokenizer::reset(std::string_view path) {
  previouslyProcessedDotOp_ = false;
  if (path.empty()) {
    index_ = 0;
    path_ = {};
    return false;
  }

  if (path[0] == ROOT || path[0] == AT) {
    index_ = 1;
    path_ = path;
    return true;
  }

  // We are parsing first token in a path that doesn't start with '$'. In this
  // case, we assume the path starts with '$.'. NOTE: This logic converts paths
  // like '.foo' into '$..foo' which uses deep scan operator (..)
  previouslyProcessedDotOp_ = true;
  index_ = 0;
  path_ = path;

  return true;
}

bool JsonPathTokenizer::hasNext() const {
  return index_ < path_.size();
}

std::optional<JsonPathTokenizer::Token> JsonPathTokenizer::getNext() {
  if (previouslyProcessedDotOp_ || match(DOT)) {
    previouslyProcessedDotOp_ = false;
    if (hasNext()) {
      if (match(DOT)) {
        // '..' Recrusive operator
        if (!hasNext() || path_[index_] == DOT) {
          // '..' Recrusive operator cannot exist by itself or followed by a dot
          return std::nullopt;
        }
        previouslyProcessedDotOp_ = true;
        return JsonPathTokenizer::Token{"", Selector::RECURSIVE};
      } else if (path_[index_] != OPEN_BRACKET) {
        return matchDotKey();
      }
    }
    // Simply ignore the '.' followed by '['. This allows non-standard paths
    // like '$.[0].[1].[2]' supported by Jayway / Presto.
  }

  if (match(OPEN_BRACKET)) {
    skipWhitespace();
    auto token = match(DOUBLE_QUOTE)
        ? matchQuotedSubscriptKey(DOUBLE_QUOTE)
        : (match(SINGLE_QUOTE) ? matchQuotedSubscriptKey(SINGLE_QUOTE)
                               : matchUnquotedSubscriptKey());
    skipWhitespace();
    if (!token || !match(CLOSE_BRACKET)) {
      return std::nullopt;
    }
    return token;
  }

  return std::nullopt;
}

bool JsonPathTokenizer::match(char expected) {
  if (hasNext() && path_[index_] == expected) {
    index_++;
    return true;
  }
  return false;
}

std::optional<JsonPathTokenizer::Token> JsonPathTokenizer::matchDotKey() {
  if (hasNext() && path_[index_] == STAR) {
    index_++;
    return JsonPathTokenizer::Token{"", Selector::WILDCARD};
  }
  auto start = index_;
  while (hasNext() && isDotKeyFormat(path_[index_])) {
    if (std::isspace(path_[index_])) {
      return std::nullopt;
    }
    index_++;
  }
  if (index_ == start) {
    return std::nullopt;
  }
  auto key = path_.substr(start, index_ - start);
  return JsonPathTokenizer::Token{std::string(key), Selector::KEY_OR_INDEX};
}

void JsonPathTokenizer::skipWhitespace() {
  while (hasNext() && std::isspace(path_[index_])) {
    index_++;
  }
}

std::optional<JsonPathTokenizer::Token>
JsonPathTokenizer::matchUnquotedSubscriptKey() {
  auto start = index_;
  while (hasNext() && isUnquotedBracketKeyFormat(path_[index_])) {
    index_++;
  }
  if (index_ == start) {
    return std::nullopt;
  }

  auto key = path_.substr(start, index_ - start);
  if (key.size() == 1 && key[0] == STAR) {
    return JsonPathTokenizer::Token{"", Selector::WILDCARD};
  }

  return JsonPathTokenizer::Token{std::string(key), Selector::KEY_OR_INDEX};
}

// Reference Presto logic in
// src/test/java/io/prestosql/operator/scalar/TestJsonExtract.java and
// src/main/java/io/prestosql/operator/scalar/JsonExtract.java
std::optional<JsonPathTokenizer::Token>
JsonPathTokenizer::matchQuotedSubscriptKey(char quote) {
  bool escaped = false;
  std::string token;
  while (hasNext() && (escaped || path_[index_] != quote)) {
    if (escaped) {
      if (path_[index_] != quote && path_[index_] != BACK_SLASH) {
        return std::nullopt;
      }
      escaped = false;
      token.append(1, path_[index_]);
    } else {
      if (path_[index_] == BACK_SLASH) {
        escaped = true;
      } else if (path_[index_] == quote) {
        return std::nullopt;
      } else {
        token.append(1, path_[index_]);
      }
    }
    index_++;
  }
  if (escaped || token.empty() || !match(quote)) {
    return std::nullopt;
  }
  return JsonPathTokenizer::Token{token, Selector::KEY};
}

static std::string SelectorToString(JsonPathTokenizer::Selector selector) {
  switch (selector) {
    case JsonPathTokenizer::Selector::KEY:
      return "KEY";
    case JsonPathTokenizer::Selector::WILDCARD:
      return "WILDCARD";
    case JsonPathTokenizer::Selector::KEY_OR_INDEX:
      return "KEY_OR_INDEX";
    default:
      return "Unknown";
  }
}
std::string JsonPathTokenizer::Token::toString() const {
  return value + " : " + SelectorToString(selector);
}

} // namespace facebook::velox::functions
